### PR TITLE
FormItem validation for analytics

### PR DESCRIPTION
**Pull request by erenbesel**
Adds a new protocol to return validation status that only required validators can conform to.
Keeps the current logic and additionally checks for the new status via `triggerValidationErrorIfNeeded()`

Makes all existing public validators spi as they were not intended to be public.

### CODE REVIEW

File: Adyen/Validators/ValidationError.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            return validationError
+        }
+    }
+}
+
+/// Protocol that defines the interface for validation errors.
+@_spi(AdyenInternal)
+public protocol ValidationError: LocalizedError {}
+
+/// Protocol created spefically for analytics related validation error.
```
Comment (atmamont) [inline Adyen/Validators/ValidationError.swift]: Let's extract this one to the Analytics module(folder)?

**Comment by atmamont** [comment]
Wow, nice that we have a spellcheck on CI!

File: AdyenCard/Components/Card/CardViewControllerItemsProvider.swift [hunk 2]
```diff
@@ -269,12 +235,31 @@ extension CardViewController {
             return item
         }()
         
-        private func triggerInfoEvent(
-            of type: AnalyticsEventInfo.InfoType,
-            target: AnalyticsEventTarget,
-            message: String? = nil
-        ) {
-            onDidTriggerEvent?(type, target)
+        private func setupEventTriggers(for item: FormTextItem, target: AnalyticsEventTarget) {
```
Comment (atmamont) [inline AdyenCard/Components/Card/CardViewControllerItemsProvider.swift:238]: 👍

File: AdyenCard/Validators/CardNumberValidator.swift [hunk 3]
```diff
@@ -38,12 +39,30 @@ public final class CardNumberValidator: Validator {
...
+        
+        if !isEnteredBrandSupported {
+            return .invalid(CardValidationError.cardUnsupported)
+        }
+        
+        if value.isEmpty {
+            return .invalid(CardValidationError.cardNumberEmpty)
+        }
+        
+        if value.count < 12 {
```
Comment (atmamont) [inline AdyenCard/Validators/CardNumberValidator.swift]: This is a magic number, let's extract it to a named constant?

**Review by atmamont — DISMISSED** [review]
👏

File: Adyen/Validators/ValidationError.swift (hunk 1 above)
Comment (erenbesel) [reply Adyen/Validators/ValidationError.swift]: 776dbfd3ef4ea752d9bdeef328c4327425240f4b

File: AdyenCard/Validators/CardNumberValidator.swift (hunk 3 above)
Comment (erenbesel) [reply AdyenCard/Validators/CardNumberValidator.swift]: good catch! 776dbfd3ef4ea752d9bdeef328c4327425240f4b

**Review by atmamont — DISMISSED** [review]

File: AdyenCard/Components/Card/CardViewControllerItemsProvider.swift [hunk 4]
```diff
@@ -144,12 +135,7 @@ extension CardViewController {
             securityCodeItem.localizationParameters = localizationParameters
             securityCodeItem.identifier = ViewIdentifierBuilder.build(scopeInstance: scope, postfix: "securityCodeItem")
             
-            securityCodeItem.onDidBeginEditing = { [weak self] in
-                self?.triggerInfoEvent(of: .focus, target: .securityCode)
-            }
-            securityCodeItem.onDidEndEditing = { [weak self] in
-                self?.triggerInfoEvent(of: .unfocus, target: .securityCode)
-            }
+            setupEventTriggers(for: securityCodeItem, target: .securityCode)
```
Comment (nauaros) [inline AdyenCard/Components/Card/CardViewControllerItemsProvider.swift:138]: Nice!

File: AdyenCard/Validators/CardValidationError.swift [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            return 920
+        case .securityCodePartial:
+            return 921
+        }
+    }
+    
+    internal var analyticsErrorMessage: String {
+        switch self {
+        case .cardNumberEmpty:
+            return "Empty card number field."
```
Comment (nauaros) [inline AdyenCard/Validators/CardValidationError.swift:70]: Are we gonna localized this error messages?

**Review by nauaros — APPROVED** [review]

File: Adyen/UI/Form/Items/Value/Validatable/FormValidatableValueItemView.swift [hunk 6]
```diff
@@ -103,9 +103,20 @@ open class FormValidatableValueItemView<ValueType, ItemType: FormValidatableValu
     
     private func hideAlertLabel(_ hidden: Bool, animated: Bool = true) {
```
Comment (goergisn) [inline Adyen/UI/Form/Items/Value/Validatable/FormValidatableValueItemView.swift:104]: Not related to the changes, but maybe by renaming `hidden` to `shouldHide` or `hide` it makes the logic easier to read :)

File: Adyen/UI/Form/Items/Value/Validatable/FormValidatableValueItemView.swift [hunk 7]
```diff
@@ -103,9 +103,20 @@ open class FormValidatableValueItemView<ValueType, ItemType: FormValidatableValu
...
         guard hidden || alertLabel.text != nil else { return }
+        if !hidden {
+            triggerValidationErrorIfNeeded()
+        }
         alertLabel.adyen.hide(animationKey: "hide_alertLabel", hidden: hidden, animated: animated)
     }
     
+    // this function can be used to populate the alert label with the correct invalid message
```
Comment (goergisn) [inline Adyen/UI/Form/Items/Value/Validatable/FormValidatableValueItemView.swift:112]: The message is not super clear to me without any context rn - maybe it's not even needed as the function name is very self explainatory

**Review by neelSharma12 — APPROVED** [review]

File: AdyenCard/Validators/CardNumberValidator.swift [hunk 8]
```diff
@@ -38,12 +40,30 @@ public final class CardNumberValidator: Validator {
...
     }
     
-    public func isValid(_ value: String) -> Bool {
-        guard isEnteredBrandSupported else { return false }
-        let minimumValidCardLength = 12
-        let isValid = value.count >= minimumValidCardLength && (!isLuhnCheckEnabled || luhnCheck(value))
+    public func validate(_ value: String) -> ValidationStatus {
+        // order of checks are important to return the correct error.
+        
+        if !isEnteredBrandSupported {
```
Comment (goergisn) [inline AdyenCard/Validators/CardNumberValidator.swift:46]: Just a thought - maybe something for a later point - but reading this feels a bit odd that we're passing a `isEnteredBrandSupported` to the CardNumberValidator which basically overrides all other validation rules.

File: Adyen/Analytics/Models/AnalyticsValidationError.swift [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+import Foundation
+
+/// Protocol created specially for analytics related validation errors.
+@_spi(AdyenInternal)
+public protocol AnalyticsValidationError: ValidationError {
+    
+    /// Code of the error for analytics.
+    var analyticsErrorCode: Int { get }
+    
+    /// Message describing the error for analytics.
```
Comment (goergisn) [inline Adyen/Analytics/Models/AnalyticsValidationError.swift:16]: Maybe good to add that they are not meant to be shown to the user

File: AdyenCard/Validators/CardValidationError.swift (hunk 5 above)
Comment (goergisn) [reply AdyenCard/Validators/CardValidationError.swift:70]: These messages are only meant for the backend and not to be shown to the user.

**Review by goergisn — APPROVED** [review]

**Pull request merged**